src/Dialect/ONNX/ONNXOpsHelper.cpp

```diff
@@ -495,12 +495,34 @@ Value normalizeConstantOp(
 }
 
 // Create a DenseElementsAttr based on the shape of type.
-DenseElementsAttr createDenseElementsAttrFromShape(
-    PatternRewriter &rewriter, Value value) {
+DenseElementsAttr createDenseElementsAttrFromShape(PatternRewriter &rewriter,
+    Value value, Attribute startAttr, Attribute endAttr) {
+  // Check that end is provided
+
   auto inType = value.getType().cast<ShapedType>();
   auto shape = inType.getShape();
-  SmallVector<int64_t, 1> dims = {inType.getRank()};
-  SmallVector<int64_t, 4> values(shape.begin(), shape.end());
+  int64_t rank = inType.getRank();
+
+  int64_t start = 0;
+  int64_t end = rank;
+
+  if (startAttr) {
+    start = startAttr.cast<IntegerAttr>().getSInt();
+  }
+  if (endAttr) {
+    end = endAttr.cast<IntegerAttr>().getSInt();
+  }
+
+  // Normalize if start/end are not in (0, ..., rank)
+  if (start < 0) {
+    start = start + rank;
+  }
+  if (end < 0) {
+    end = end + rank;
+  }
+
```

Comment:
would it make sense to add asserts to make sure that start and end are now in inclusively 0..rank-1 for start, 0..rank for end?

Reply:
Good catch! This was actually wrong because the behaviour is meant to do be the following

> "Optional attributes start and end can be used to compute a slice of the input tensor's shape."
  "If start axis is omitted, the slice starts from axis 0."
  "The end axis, if specified, is exclusive (and the returned value will not include the size of that axis)."
  "If the end axis is omitted, the axes upto the last one will be included."
  "Negative axes indicate counting back from the last axis."
  "Note that axes will be clipped to the range [0, r-1], where r is the"
  "rank of the input tensor if they are out-of-range (after adding r in the case of"
  "negative axis). Thus, specifying any end value > r is equivalent to specifying an end"
  "value of r, and specifying any start value < -r is equivalent to specifying a start"
  "value of 0."
  
  I've rejigged the logic to use `getDataShapeBounds` which handles the clipping logic. This puts the normalizing behavior in one place, where it is handled correctly